### PR TITLE
C# 12: Support for collection expressions.

### PR DESCRIPTION
**Pull request by michaelnebel**
In this PR we introduce extractor, QL library and dataflow support for *collection* *expressions*.
Example of collection expressions are `[1, x, 3]` and `[2, .. y]` in
``` csharp
var x = 6;
int[] y = [1, x, 3];
int[] z = [2, .. y];
```
Microsoft has documented collection expressions [here](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/collection-expressions). However, it is worth noting that even though the documentation states that collection expressions are also fully supported for inline arrays - but this appears not to be the case - which is indicated [here](https://github.com/dotnet/docs/issues/38916).

### CODE REVIEW

**Comment by michaelnebel** [comment]
DCA
(1) Performance looks good on average.
(2) A lot of extraction errors have been removed from the MS repositories (it looks like the collection expression feature has been used quite a lot already). There is some discrepancy in the alerts - especially in the repositories that uses the collection expression features. I still need to look into some of these.

File: csharp/ql/lib/semmle/code/csharp/exprs/Expr.qll [hunk 1]
```diff
@@ -1216,3 +1216,47 @@ class WithExpr extends Expr, @with_expr {
...
+   */
+  Expr getAnElement() { result = this.getElement(_) }
+
+  override Type getType() {
+    super.getType() instanceof NullType and
+    exists(CastExpr ce | ce.getExpr() = this and result = ce.getType())
+    or
+    not super.getType() instanceof NullType and
+    result = super.getType()
+  }
```
Comment (tamasvajk) [inline csharp/ql/lib/semmle/code/csharp/exprs/Expr.qll:1257]: When would the collection literal have null type? Is this related to `CollectionBuilderAttribute`?

Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/exprs/Expr.qll:1257]: If there is an explicit cast applied to the collection expression, then it doesn't carry a typesymbol with its type information. So something like
```
int[] x = (int [])[1,2,3];
```

Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/exprs/Expr.qll:1257]: This is the only case where I have observed that the collection expression doesn't carry a type symbol (according to the documentation there are clear rules on, how the collection expression is typed if there are ambiguities and in those case it does carry a symbol with the type information - but not in the case with the explicit cast).

**Review by tamasvajk — APPROVED** [review]
LGTM. The DB stats seem to be manually added. Should the stats be generated to have real numbers?

**Comment by michaelnebel** [comment]
> LGTM. The DB stats seem to be manually added. Should the stats be generated to have real numbers?

Yes, they are manually added.
@hvitved  is in the process of figuring out whether we need the DB stats at all. I will postpone updating stats unless we run into performance issues, or if it turns out that they can't be scrapped all together or if I manage to finish all the C# 12 updates before we identify, whether stats can be scrapped :-D

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 2]
```diff
@@ -250,6 +250,10 @@ module LocalFlow {
         scope = e2 and
         isSuccessor = true
         or
+        e1 = e2.(CollectionExpression).getAnElement().(SpreadElementExpr).getExpr() and
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:253]: I think it is better to break this up into two steps:

For `[..x]`, add a local flow step from `x` to `..x`, and another local flow step from `..x` to `[..x]`. We can then implement `expectsContent` on the `..x` node, so that flow will only be allowed when we are in fact tracking flow inside a collection.

**Review by hvitved — COMMENTED** [review]
LGTM. One suggestion.

**Comment by michaelnebel** [comment]
> DCA (1) Performance looks good on average. (2) A lot of extraction errors have been removed from the MS repositories (it looks like the collection expression feature has been used quite a lot already). There is some discrepancy in the alerts - especially in the repositories that uses the collection expression features. I still need to look into some of these.

I did some spot checking on the alerts changes on `dotnet/roslyn`.
- All removed `cs/unused-collection` were false positives.
- Checked several `cs/useless-assignment-of-local` and these were false positives as well as the query couldn’t detect uses of variables in collection expressions.
- The added alerts to `cs/call-to-object-tostring` are true positives.

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll (hunk 2 above)
Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:253]: Wow, that is a really nice catch @hvitved !!!

**Comment by michaelnebel** [comment]
DCA looks good.

**Review by hvitved — APPROVED** [review]
